### ISSUE TITLE
sub-resources return 405 for PUT/DELETE or not called at all

### ISSUE THREAD

**Issue by justjacksonn** [mulesoft/raml-for-jax-rs#107]
Hi,

I am testing out the ability to have a 3-level deep API in RAML turn to JAX-RS and deployed. I have get/put/post/delete methods for the first/top level of the endpoint. Those all seem to work. So something like PUT /endpoint called my implemented PUT method. The PUT call with an entity works fine, as does POST, GET, DELETE. When I define a sub-resource in my raml:

/endpoint:
  put:
  post:
  etc...
  /{subResource}:
    put:
      ..stuff

For some reason the PUT request to this sub-resource returns with a 405 method not allowed. The response headers show a GET,POST,OPTIONS for the Allow header. DELETE also fails at this level.

I would have thought maybe a config issue with Jetty blocking PUT/DELETE, but the fact that it works at the first endpoint level seems to me that it's allowed through. I can't find anything via basic google searching around enabled PUT/DELETE other than generic info in general. Nothing that would indicate PUT/DELETE doesn't work for nested paths.

I am also seeing that even GET doesn't work on a 3rd level sub-resource.. the generated interface shows something like:

@Path("{sub}/{sub2}")
@GET
@Produces({"application/json"})
....

I implement the interface with the @overide as I do the others, and it never gets called. I've tried without any accept/content-type, with each, with both, nothing calls the method.

So again I am a little bewildered how all method types work with the first level of resource, the 2nd level only works with get/post, and the third doesn't work at all.

I have yet to try with a different container as that is not an option for me, but I assume others have deployed to Jetty as it's pretty popular. In my case I am using Jetty 9.2. I may try to get an older version.. 9.1 or something and see if that changes anything.

**Comment by ddossot** [comment]
Can you try a test JAX-RS annotated class, not generated by RAML? If it works, it will be interesting to know the difference between the generated code and the code you wrote.

**Comment by justjacksonn** [comment]
Ok..first, I feel stupid for not thinking of that myself! Great idea! So I did two things.. first I created just a class with the @Path.. resources for GET, PUT, POST and DELETE for both the top level and sub-resource level. ALL calls worked. So then I figured, I should build an interface with the @Path, @GET, etc in it, like how RAML-JAXRS does it, put that in a different package, then implement the interface with just @Overide methods like how I am doing it now for the generated JAX-RS code. ALL calls work! 

I ALSO tested NOT putting the @Path() in my implementation class, it is in the interface class. That DID NOT work. @ddossot you were saying you did NOT put the @Path in your implementation class, and that your deployment worked. I am using Java 8, Jetty 9.2, not sure why, but without the @Path in the implementation class, it doesn't find the resources at all within the implementation class despite that it implements an interface that declares all of that. So I am unsure why/how your specific setup works.

Back to the issue though.. so it seems (thanks to your idea) that PUT/DELETE DO work, just not with the generated RAML-JAXRS code. Now the question is..what/why that is happening. I will continue to try to figure this out..but may the authors of the raml-jaxrs code generator can play around with this and see what they come up with as well.

**Comment by ddossot** [comment]
Thanks for further testing, and please do not feel stupid, we're in this together :wink: 

From the JAX-RS 2.0 spec:

> ## 3.6 Annotation Inheritance
> JAX-RS annotations may be used on the methods and method parameters of a super-class or an implemented interface. Such annotations are inherited by a corresponding sub-class or implementation class method provided that the method and its parameters do not have any JAX-RS annotations of their own. Annotations on a super-class take precedence over those on an implemented interface. The precedence over conflicting annotations defined in multiple implemented interfaces is implementation specific. Note that inheritance of class or interface annotations is not supported.
> If a subclass or implementation method has any JAX-RS annotations then all of the annotations on the superclass or interface method are ignored.

So the question is: are the methods in your implementation class JAX-RS-annotated?

**Comment by justjacksonn** [comment]
The methods in my implementation class are NOT JAX-RS annotated. Only the interface is. I basically use the IDE (Eclipse and IntelliJ both support this) to auto-implement the interface stubs. They properly provide the implementations with @Overide on the method, and that's it. The @PathParam() interface definitions show up as just String <name> and so forth in the implementing class.

**Comment by justjacksonn** [comment]
SON OF A!!! Man, I hate when I check, double check..test.. and after all that and all this, I go back again and find an error.. on my part! So.. after you posted that bit 3.6 above.. I looked again at my methods. Turns out, I had @Overide... which was correct, but I also had @PathParam() in the method signature of my method. By removing that.. now it works! 

AAhh. ok.. I figured it out. IntelliJ apparently DOES add the @PathParam() when I use the Generate -> Implement Methods.. and here I thought it was as good as Eclipse ;).

SO....my apologies all for making something out of nothing. Sub resources, sub-sub resources.. all work fine. Just need to make sure using the IDE to generate unimplemented methods doesn't add any JAX-RS annotations to the generated methods is all.

Closing.
